src/NaNMath.jl

```diff
@@ -22,7 +22,9 @@ sqrt(x::Real) = x < 0.0 ? NaN : Base.sqrt(x)
 # Don't override built-in ^ operator
 pow(x::Float64, y::Float64) = ccall((:pow,libm),  Float64, (Float64,Float64), x, y)
 pow(x::Float32, y::Float32) = ccall((:powf,libm), Float32, (Float32,Float32), x, y)
-pow(x::Number,y::Number) = pow(float(x),float(y))
+# Need `promote` here, otherwise we'll end up with an infinite recursion
+# in the case of `x` and `y` being different floating types.
+pow(x::Number, y::Number) = pow(promote(float(x), float(y))...)
```

Comment:
This seems not ideal (see https://github.com/JuliaDiff/DiffRules.jl/pull/55#discussion_r794037403), it would be better to use
```suggestion
pow(x::Number, y::Number) = pow(map(float, promote(x, y))...)
```